pkg/controller/attach_detach.go

```diff
@@ -157,7 +163,38 @@ func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, azVolumeAttac
 		cloudCtx, cloudCancel := context.WithTimeout(context.Background(), cloudTimeout)
 		defer cloudCancel()
 		updateCtx := context.Background()
+		// attempt to attach the disk to a node
 		response, attachErr := r.attachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
+		// if the disk is attached to a different node
+		if _, ok := attachErr.(*volerr.DanglingAttachError); ok {
+			// get disk, node and current attachment name
+			diskName, err := azureutils.GetDiskName(azVolumeAttachment.Spec.VolumeID)
+			if err != nil {
+				klog.Warningf("failed to extract disk name from volumeID (%s): %v", azVolumeAttachment.Spec.VolumeID, err)
+			}
+			currentNodeName := string(attachErr.(*volerr.DanglingAttachError).CurrentNode)
+			currentAttachmentName := azureutils.GetAzVolumeAttachmentName(diskName, currentNodeName)
+
+			// Check if AzVolumeAttachment for the Dangling attachment exists
+			_, err = r.azVolumeClient.DiskV1alpha1().AzVolumeAttachments(r.namespace).Get(cloudCtx, currentAttachmentName, metav1.GetOptions{})
+			if errors.IsNotFound(err) {
+				// AzVolumeAttachment doesn't exist so we only need to detach disk from cloud
+				err = r.cloudDiskAttacher.UnpublishVolume(ctx, azVolumeAttachment.Spec.VolumeID, currentNodeName)
+				if err != nil {
+					klog.Errorf("failed to detach volume (%s) from node (%s). Error: %v", azVolumeAttachment.Spec.VolumeID, currentNodeName, err)
+				}
+			} else {
+				// AzVolumeAttachment exist so we need to detach disk through crdProvisioner
+				err = r.crdDetacher.UnpublishVolume(ctx, azVolumeAttachment.Spec.VolumeID, currentNodeName, make(map[string]string))
+				if err != nil {
+					klog.Errorf("failed to unpublish AzVolumeAttachment for volume (%s) and node (%s). Error: %v", azVolumeAttachment.Spec.VolumeID, currentNodeName, err)
+				}
+				klog.Errorf("failed to get AzVolumeAttachment for (%s) Err: %v", azVolumeAttachment.Name, err)
+			}
```

Comment:
If the attempt to detach the disk from the other node fails, we should return an error instead of trying to attach. The question is which error?

The code currently returns `status.Errorf(codes.Internal, "Could not detach volume %q from node %q: %v", diskURI, derr.CurrentNode, err)`, so we could go with that though it seems odd that a `ControllerPublishVolume` call would try to detach a disk from another node. We could simply return the `DanglingAttachError` itself which actually has a clearer error message:

```go
			attachErr := fmt.Sprintf(
				"disk(%s) already attached to node(%s), could not be attached to node(%s)",
				diskURI, *disk.ManagedBy, nodeName)
```

Honestly, I think it'd be best to simply return the dangling attach error here.

Reply:
attachErr is what eventually gets reflected in the `AzVolumeAttachment` CRI status. I agree though that there is no point in retrying the attachment second time if detach failed.